cpp/src/arrow/flight/internal.cc

```diff
@@ -49,27 +50,153 @@ Status FromGrpcStatus(const grpc::Status& grpc_status) {
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::Cancelled);
+      break;
+    case grpc::StatusCode::INVALID_ARGUMENT:
+      code = StatusCode::Invalid;
+      message =
+          util::StringBuilder("gRPC returned invalid argument error, with message: ",
+                              grpc_status.error_message());
+      break;
+    case grpc::StatusCode::DEADLINE_EXCEEDED:
+      code = StatusCode::IOError;
+      message =
+          util::StringBuilder("gRPC returned deadline exceeded error, with message: ",
+                              grpc_status.error_message());
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::TimedOut);
+      break;
+    case grpc::StatusCode::NOT_FOUND:
+      code = StatusCode::IOError;
+      message = util::StringBuilder("gRPC returned not found error, with message: ",
+                                    grpc_status.error_message());
+      break;
+    case grpc::StatusCode::ALREADY_EXISTS:
+      code = StatusCode::Invalid;
+      message = util::StringBuilder("gRPC returned already exists error, with message: ",
+                                    grpc_status.error_message());
+      break;
+    case grpc::StatusCode::PERMISSION_DENIED:
+      code = StatusCode::IOError;
+      message =
+          util::StringBuilder("gRPC returned permission denied error, with message: ",
+                              grpc_status.error_message());
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::Unauthorized);
+      break;
+    case grpc::StatusCode::RESOURCE_EXHAUSTED:
+      code = StatusCode::Invalid;
+      message =
+          util::StringBuilder("gRPC returned resource exhausted error, with message: ",
+                              grpc_status.error_message());
+      break;
+    case grpc::StatusCode::FAILED_PRECONDITION:
+      code = StatusCode::Invalid;
+      message =
+          util::StringBuilder("gRPC returned precondition failed error, with message: ",
+                              grpc_status.error_message());
+      break;
+    case grpc::StatusCode::ABORTED:
+      code = StatusCode::IOError;
+      message = util::StringBuilder("gRPC returned aborted error, with message: ",
+                                    grpc_status.error_message());
+      detail = std::make_shared<FlightStatusDetail>(FlightStatusCode::Internal);
+      break;
+    case grpc::StatusCode::OUT_OF_RANGE:
+      code = StatusCode::Invalid;
+      message = util::StringBuilder("gRPC returned out-of-range error, with message: ",
+                                    grpc_status.error_message());
+      break;
+    case grpc::StatusCode::UNIMPLEMENTED:
+      code = StatusCode::NotImplemented;
+      message = util::StringBuilder("gRPC returned unimplemented error, with message: ",
+                                    grpc_status.error_message());
+      break;
+    case grpc::StatusCode::INTERNAL:
+      code = StatusCode::IOError;
```

Comment:
Unknown might be better?

Reply:
Agreed, changed. I also added a missing branch for gRPC's UNKNOWN.